### PR TITLE
bump hf deps

### PR DESCRIPTION
**Pull request by winglian**
## Summary by CodeRabbit

- **New Features**
  - Added a test to verify LoRA packing with packed datasets, ensuring correct data loader and sampler behavior.
  - Added a multi-GPU test for GRPO training with LoRA and sequence parallelism using vLLM integration.
  - Extended VLLM server to support reasoning features via new CLI arguments and worker logic.
- **Bug Fixes**
  - Updated regular expressions for LoRA target modules across multiple configuration files and documentation to match the correct model hierarchy.
  - Adjusted logic for setting batch packing options in training arguments for more predictable behavior.
  - Prevented appending null reference models during sequence parallel training.
- **Refactor**
  - Unified and streamlined data loader creation and sampler management for training and evaluation.
  - Improved state dict loading for sharded FSDP2 models with enhanced dtype and contiguity handling.
  - Removed legacy monkeypatches for Gemma3 and Mllama Flash Attention 2, simplifying patch management.
  - Simplified LoRA kernel patching logic to better handle model attribute hierarchies.
  - Moved sequence parallel group initialization to training start for GRPO trainer.
- **Chores**
  - Upgraded several Python dependencies and the `modal` package version in CI workflows.
  - Updated installation instructions for the `cut-cross-entropy` package.
  - Removed legacy and unused monkeypatch files and related documentation entries.
- **Tests**
  - Increased loss threshold in knowledge distillation tests.
  - Restricted optimizer tests to run only with supported PyTorch versions.

### CODE REVIEW

**Comment by salmanmohammadi** [comment]
I also think [this file](https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/monkeypatch/accelerate/fsdp2.py) can be updated. Let me know if you'd like a hand with this.

**Comment by NanoCode012** [comment]
@SalmanMohammadi , what needs to be changed in the linked file?

The current error seems to be due to our linked test loads from https://huggingface.co/JackFram/llama-68m which is pytorch bin model https://github.com/axolotl-ai-cloud/axolotl/actions/runs/15346236028/job/43186938948?pr=2735 (Is this our sign to drop py2.5 support?)

File: src/axolotl/monkeypatch/accelerate/fsdp2.py [hunk 1]
```diff
@@ -18,27 +18,65 @@ def fsdp2_load_full_state_dict(accelerator, model: torch.nn.Module, full_sd: dic
...
             dist.broadcast(full_param, src=0, group=mesh.get_group())
             sharded_tensor = distribute_tensor(
                 full_param, mesh, sharded_sd[param_name].placements
             )
+            to_contiguous, casting_dtype = _infer_parameter_dtype(
+                model,
+                param_name,
+                full_param,
+            )
+            sharded_tensor = _cast_and_contiguous(
```
Comment (salmanmohammadi) [inline src/axolotl/monkeypatch/accelerate/fsdp2.py:75]: How come we need the additional call to `.contiguous`?

Comment (winglian) [reply src/axolotl/monkeypatch/accelerate/fsdp2.py:75]: This is copied from upstream and modified to be similar to our previous patch

File: src/axolotl/core/trainers/grpo/trainer.py [hunk 2]
```diff
@@ -136,13 +136,22 @@ def __init__(
...
+    def train(self, *args, **kwargs):
         # Initialize the SP group
         self.sp_group = get_ring_attn_group()
         self.rank = dist.get_rank()
         self.world_size = dist.get_world_size()
         self.local_rank = dist.get_rank(group=self.sp_group)
         self.local_world_size = dist.get_world_size(group=self.sp_group)
 
+        return super().train(*args, **kwargs)
+
```
Comment (winglian) [inline src/axolotl/core/trainers/grpo/trainer.py:154]: @djsaunde Does this seem sane? we talked about this the other day since we don't have ring attention groups initialized until the train is called with the context manager.

File: requirements.txt [hunk 3]
```diff
@@ -6,20 +6,20 @@ triton>=3.0.0
...
 tokenizers>=0.21.1
-accelerate==1.6.0
-datasets==3.5.1
-deepspeed>=0.15.4
-trl==0.17.0
-hf_xet==1.1.0
+accelerate==1.7.0
+datasets==3.6.0
+deepspeed>=0.17.0
+trl==0.18.0
```
Comment (salmanmohammadi) [inline requirements.txt]: TRL released a patch https://github.com/huggingface/trl/releases/tag/v0.18.1 but not sure if there's anything that we depend on

**Review by salmanmohammadi — APPROVED** [review]
LGTM, @djsaunde could you take a pass?

File: src/axolotl/core/trainers/grpo/trainer.py (hunk 2 above)
Comment (djsaunde) [reply src/axolotl/core/trainers/grpo/trainer.py:154]: Yes, seems fine, though you don't need to re-set `self.rank` or `self.world_size`.

File: src/axolotl/core/trainers/grpo/args.py [hunk 4]
```diff
@@ -12,3 +12,5 @@
 @dataclass
 class AxolotlGRPOConfig(AxolotlTrainingMixins, GRPOConfig):
     """Axolotl GRPO Config for GRPO training"""
+
+    sequence_parallel_degree: int | None = None
```
Comment (djsaunde) [inline src/axolotl/core/trainers/grpo/args.py:16]: Why do we need this? AFAICT this isn't used

**Review by djsaunde — APPROVED** [review]
I wonder if there's an easier way to update our custom dataloader logic, seems annoying to have to update whenever upstream changes

File: src/axolotl/core/trainers/grpo/args.py (hunk 4 above)
Comment (winglian) [reply src/axolotl/core/trainers/grpo/args.py:16]: we have this code in `AxolotlGRPOSequenceParallelTrainer`:
```
num_sp_groups = num_processes // self.args.sequence_parallel_degree
```
